src/relay/backend/vm/compiler.cc

```diff
@@ -418,29 +383,194 @@ class VMFunctionCompiler : ExprFunctor<void(const Expr& expr)> {
     this->last_register_ = true_register;
   }
 
-  Instruction AllocTensorFromType(const TensorTypeNode* ttype) {
-    TVMType dltype = Type2TVMType(ttype->dtype);
-    auto tensor_type = GetRef<TensorType>(ttype);
+  Index EmitGetShape(const TensorTypeNode* ttype, Index reg) {
+    bool const_shape = true;
     std::vector<int64_t> shape;
-    for (auto dim : tensor_type->shape) {
-      shape.push_back(Downcast<tvm::Integer>(dim)->value);
+    for (auto dim : ttype->shape) {
+      if (auto kdim = dim.as<IntImm>()) {
+        shape.push_back(kdim->value);
+      } else {
+        const_shape = false;
+      }
+    }
+    if (const_shape) {
+      int64_t s = shape.size();
+      DLContext cpu_ctx;
+      cpu_ctx.device_type = kDLCPU;
+      cpu_ctx.device_id = 0;
+      auto shape_tensor = NDArray::Empty({s}, Type2TVMType(Int(64)), cpu_ctx);
+      int64_t* dims = reinterpret_cast<int64_t*>(shape_tensor->data);
+      for (size_t i = 0; i < shape.size(); ++i) {
+        dims[i] = shape[i];
+      }
+      size_t konst_idx = context_->constants.size();
+      context_->constants.push_back(shape_tensor);
+      Emit(Instruction::LoadConst(konst_idx, NewRegister()));
+      return last_register_;
     }
-    return Instruction::AllocTensor(shape, dltype, NewRegister());
+    // For dynamic shape, we need insert shape_of op to get its shape at runtime
+    auto attrs = make_node<ShapeOfAttrs>();
+    attrs->dtype = Int(64);
+    static const Op& op = Op::Get("shape_of");
+    auto input = VarNode::make("input", GetRef<Type>(ttype));
+    auto expr = CallNode::make(op, {input}, Attrs(attrs), {});
+    auto func = FunctionNode::make({input}, expr, IncompleteTypeNode::make(Kind::kType), {});
+    auto mod = ModuleNode::make({}, {});
+    auto main_gv = GlobalVarNode::make("main");
+    mod->Add(main_gv, func);
```

Comment:
The reason for us to create a `mod` here is to infer type, right?

Reply:
yes